Cabal/Distribution/Parsec/Types/ParseResult.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE CPP #-}
+module Distribution.Parsec.Types.ParseResult (
+    ParseResult,
+    runParseResult,
+    recoverWith,
+    parseWarning,
+    parseFailure,
+    parseFatalFailure,
+    parseFatalFailure',
+    parseWarnings',
+    ) where
+
+import           Distribution.Compat.Prelude
+import           Prelude ()
+
+import           Distribution.Parsec.Types.Common
+                 (PError (..), PWarnType (..), PWarning (..), Position (..))
+
+-- | A monad with failure and accumulating errors
+newtype ParseResult a = PR { runPR :: PRState -> (Maybe a, PRState) }
+
+data PRState = PRState ![PWarning] ![PError]
+
+emptyPRState :: PRState
+emptyPRState = PRState [] []
+
+runParseResult :: ParseResult a -> ([PWarning], [PError], Maybe a)
```

Comment:
-- | Destruct a 'ParseResult' into the emitted warnings and errors, and possibly the final result if there were no errors.